### ISSUE TITLE
Special reference `__class__` is considered undefined in nested lexical scope.

### ISSUE THREAD

**Issue by emcd** [PyCQA/pylint#4032]
### Reproducer

```python
class Something:
    def _enclose( ):
        def __init__( self ):
            print( __class__ )
        return __init__
    __init__ = _enclose( )

something = Something( )
```

### Steps to reproduce
1. Run `pylint --score=no --reports=no --enable=all --disable=missing-module-docstring,missing-class-docstring,multiple-statements,too-few-public-methods,no-method-argument,unused-argument` on the reproducer above.
2. Note that Pylint complains about `__class__` being an undefined variable.
3. Run the reproducer in a Python interpreter and note that it executes correctly.

### Current behavior
```
$ pylint --score=no --reports=no --enable=all --disable=missing-module-docstring,missing-class-docstring,multiple-statements,too-few-public-methods,no-method-argument,unused-argument pylint-bugs/nested_dunder_class.py 
************* Module nested_dunder_class
pylint-bugs/nested_dunder_class.py   4,19 [undefined-variable] Undefined variable '__class__'
```
```
$ python3 pylint-bugs/nested_dunder_class.py
<class '__main__.Something'>
$
```

### Expected behavior
No error since this is valid Python.

### pylint --version output
```
$ pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.6.12 (default, Nov  8 2020, 15:40:05) 
[GCC 5.4.0 20160609]
```
